engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -320,6 +340,44 @@ long hashCodeForArray(
     }
   }
 
+  /**
+   * Two maps are considered equal, if they have the same entries. Note that we do not care about
+   * ordering.
+   */
+  @Specialization(guards = "interop.hasHashEntries(selfMap)")
+  long hashCodeForMap(
+      Object selfMap,
+      @CachedLibrary(limit = "5") InteropLibrary interop,
+      @Cached HashCodeAnyNode hashCodeNode) {
+    Object[] keys;
+    Object[] values;
+    int mapSize;
+    try {
+      mapSize = (int) interop.getHashSize(selfMap);
+      keys = new Object[mapSize];
+      values = new Object[mapSize];
+      Object entriesIterator = interop.getHashEntriesIterator(selfMap);
+      int arrIdx = 0;
+      while (interop.hasIteratorNextElement(entriesIterator)) {
+        Object entry = interop.getIteratorNextElement(entriesIterator);
+        keys[arrIdx] = interop.readArrayElement(entry, 0);
+        values[arrIdx] = interop.readArrayElement(entry, 1);
+        arrIdx++;
+      }
+    } catch (UnsupportedMessageException | StopIterationException | InvalidArrayIndexException e) {
+      throw new IllegalStateException(e);
+    }
+    assert keys.length == values.length;
+    // We don't care about the order of keys and values, so we just sum all their hash codes.
+    long keysHashCode = 0;
+    long valuesHashCode = 0;
+    for (int i = 0; i < keys.length; i++) {
+      keysHashCode += hashCodeNode.execute(keys[i]);
+      valuesHashCode += hashCodeNode.execute(values[i]);
+    }
+    return Arrays.hashCode(new long[] {keysHashCode, valuesHashCode, mapSize});
```

Comment:
Okay maybe this is a nitpick, but if we didn't split the process into 2 stages of `read keys/values into array` and `hash them`, but did all in one go, we could avoid allocating these arrays.

What do you think? I don't want to do micro optimizations here, but it seems like a rather cheap win and the code should be +- as readable as it is now.

Reply:
> What do you think? I don't want to do micro optimizations here, but it seems like a rather cheap win and the code should be +- as readable as it is now.

I am definitely making this opt. Shorter and faster code. Thanks for spotting this.